h2o-core/src/main/java/water/parser/ParseDataset.java

```diff
@@ -774,7 +774,16 @@ private FVecParseWriter makeDout(ParseSetup localSetup, int chunkOff, int nchunk
       byte[] zips = vec.getFirstBytes();
       ZipUtil.Compression cpr = ZipUtil.guessCompressionMethod(zips);
       if (localSetup._check_header == ParseSetup.HAS_HEADER) { //check for header on local file
-        byte[] bits = decryptionTool.decryptFirstBytes(ZipUtil.unzipBytes(zips, cpr, localSetup._chunk_size));
+        byte[] bits = zips;
+        if (localSetup._firstLineLen > localSetup._chunk_size) {
+          // If length of first line is known (from guess phase) and chunk size is smaller,
+          // there is a possibility of not being able to load enough data to check the whole header.
+          // Therefore, more bytes are loaded here.
+          bits = extractHeaderBytes(localSetup, decryptionTool, vec);
+        } else {
+          bits = decryptionTool.decryptFirstBytes(ZipUtil.unzipBytes(bits, cpr, localSetup._chunk_size));
```

Comment:
The compression does not seem to be required to be known at this point. We could avoid fetching first bytes from the vec twice, but this is a low-cost operation, thus kept in place right now.